src/main/java/rx/internal/producers/QueuedValueProducer.java

```diff
@@ -0,0 +1,138 @@
+     * as the queue implementation.
+     * @param child the target child subscriber
+     */
+    public QueuedValueProducer(Subscriber<? super T> child) {
+        this(child, UnsafeAccess.isUnsafeAvailable() 
+                ? new SpscLinkedQueue<Object>() : new SpscLinkedAtomicQueue<Object>());
+    }
+    /**
+     * Constructs an instance with the target child subscriber and a custom queue implementation
+     * @param child the target child subscriber
+     * @param queue the queue to use
+     */
+    public QueuedValueProducer(Subscriber<? super T> child, Queue<Object> queue) {
+        this.child = child;
+        this.queue = queue;
+        this.wip = new AtomicInteger();
+    }
+     
+    @Override
+    public void request(long n) {
+        if (n < 0) {
+            throw new IllegalArgumentException("n >= 0 required");
+        }
+        if (n > 0) {
+            BackpressureUtils.getAndAddRequest(this, n);
+            drain();
+        }
+    }
+    
+    /**
+     * Offers a value to this producer and tries to emit any queud values
+     * if the child requests allow it.
+     * @param value the value to enqueue and attempt to drain
+     * @return true if the queue accepted the offer, false otherwise
+     */
+    public boolean offer(T value) {
+        if (value == null) {
+            if (!queue.offer(NULL_SENTINEL)) {
+                return false;
+            }
+        } else {
+            if (!queue.offer(value)) {
+                return false;
+            }
+        }
+        drain();
+        return true;
+    }
+    
+    private void drain() {
+        if (wip.getAndIncrement() == 0) {
+            final Subscriber<? super T> c = child;
+            final Queue<Object> q = queue;
+            do {
+                if (c.isUnsubscribed()) {
+                    return;
+                }
+ 
+                wip.lazySet(1);
+                 
+                long r = get();
+                long e = 0;
+                Object v;
+                 
+                while (r != 0 && (v = q.poll()) != null) {
+                    try {
+                        if (v == NULL_SENTINEL) {
+                            c.onNext(null);
+                        } else {
+                            @SuppressWarnings("unchecked")
+                            T t = (T)v;
+                            c.onNext(t);
+                        }
+                    } catch (Throwable ex) {
+                        Exceptions.throwIfFatal(ex);
+                        Throwable ex1 = OnErrorThrowable.addValueAsLastCause(ex, v != NULL_SENTINEL ? v : null);
+                        c.onError(ex1);
+                        return;
+                    }
+                    if (c.isUnsubscribed()) {
+                        return;
+                    }
+                    r--;
+                    e++;
+                }
+                 
+                if (e != 0) {
```

Comment:
Because requests and available values run independently and you need to come back for more values even if get() == Long.MAX_VALUE and you'd need to remember you are on a fast-path anyway. I'll update the ```if```.